src/server/conn.rs

```diff
@@ -114,45 +123,102 @@ impl Conn {
         Ok(())
     }
 
-    pub fn read<T, S>(&mut self, _: &mut EventLoop<Server<T, S>>) -> Result<Vec<ConnData>>
+
+    pub fn on_readable<T, S>(&mut self,
+                             event_loop: &mut EventLoop<Server<T, S>>)
+                             -> Result<Vec<ConnData>>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        match self.conn_type {
+            ConnType::HandShake => self.handshake(event_loop),
+            ConnType::Rpc => self.read_rpc(event_loop),
+            ConnType::Snapshot => self.read_snapshot(event_loop),
+        }
+    }
+
+    fn handshake<T, S>(&mut self, event_loop: &mut EventLoop<Server<T, S>>) -> Result<Vec<ConnData>>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        let msg_or_none = try!(self.read_one_message());
+        if msg_or_none.is_none() {
+            return Ok(vec![]);
+        }
+
+        let msg = msg_or_none.unwrap();
+        match msg.get_msg_type() {
+            MessageType::None => {
```

Comment:
Yes, but we can do it later. 
That's a minimal change and don't involve kvproto.